File: LeetCode/165. Compare Version Numbers/index165.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {
	fmt.Println(compareVersion("0.1", "1.1"))       // -1
	fmt.Println(compareVersion("1.0.1", "1"))       // 1
	fmt.Println(compareVersion("7.5.2.4", "7.5.3")) //-1
	fmt.Println(compareVersion(".1", "0.001"))      //0
	fmt.Println(compareVersion("1.01", "1.001"))    //0
	fmt.Println(compareVersion("1.0", "1.0.0"))     //0
}

func compareVersion(version1 string, version2 string) int {
	v1 := strings.Split(version1, ".")
	v2 := strings.Split(version2, ".")

	swaped := 1
	if len(v2) > len(v1) {
		v1, v2 = v2, v1
		swaped = -1
	}

	for i := 0; i < len(v1); i++ {
		i1, _ := strconv.Atoi(v1[i])
		if i >= len(v2) {
			if i1 > 0 {
				return swaped
			}
		} else {
			i2, _ := strconv.Atoi(v2[i])
			if i1 > i2 {
				return swaped
			} else if i1 < i2 {
				return swaped * -1
			}
		}
	}
	return 0
}

File: LeetCode/165. Compare Version Numbers/index165.js

/**
 * @param {string} version1
 * @param {string} version2
 * @return {number}
 */
 var compareVersion = function (version1, version2) {
    let r1 = r2 = 0
    let l1 = l2 = 0
    let n1 = n2 = 0
    while (l1 < version1.length || l2 < version2.length) {
        n1 = 0
        n2 = 0
        while (l1 < version1.length && r1 < version1.length && version1[r1] !== ".") {
            r1++
        }
        n1 = ~~version1.slice(l1, r1)
        l1 = r1 + 1
        r1++
        while (l2 < version2.length && r2 < version2.length && version2[r2] !== ".") {
            r2++
        }
        n2 = ~~version2.slice(l2, r2)
        l2 = r2 + 1
        r2++
        if (n1 !== n2) {
            return n1 > n2 ? 1 : -1
        }
    }

    return 0
};

console.log(compareVersion("0.1", "1.1"))       // -1
console.log(compareVersion("1.0.1", "1"))       // 1
console.log(compareVersion("7.5.2.4", "7.5.3")) //-1
console.log(compareVersion(".1", "0.001"))      //0
console.log(compareVersion("1.01", "1.001"))    //0
console.log(compareVersion("1.0", "1.0.0"))     //0
